### ISSUE TITLE
SerializationException when trying to create an AAD application

### ISSUE THREAD

**Issue by estiller** [Azure/azure-sdk-for-net#1863]
Hi,
I'm trying to use the Azure Graph RBAC SDK (Microsoft.Azure.Graph.RBAC) to create an application on AAD, however I get a serialization exception.
The full code to reproduce the issue can be found [here](https://github.com/estiller/GraphRBACException). The thing about this is that while an exception is thrown, the application is indeed created.

Below is the main piece of the code which demonstrates the problem:

``` cs
static string CreateApplication(string userToken)
{
    var appInfo = new ApplicationCreateParameters
    {
        DisplayName = "TestApp",
        AvailableToOtherTenants = false,
        Homepage = "http://mydomain.com/",
        IdentifierUris = new List<string> {"http://www.mydomain.com/" + Guid.NewGuid()},
        PasswordCredentials = new List<PasswordCredential>
        {
            new PasswordCredential(
                keyId: Guid.NewGuid().ToString(),
                value: "1234",
                startDate: DateTime.Now,
                endDate: DateTime.Now.AddYears(1))
        }
    };

    using (var client = CreateManagementClient(userToken))
    {
        try
        {
            var app = client.Application.Create(appInfo); // Throws a Serialization Exception, but the application is created on AAD
            return app.AppId;
        }
        catch (SerializationException ex)  // This is a workaround - the application was created!
        {
            var jsonResponse = (JObject)JsonConvert.DeserializeObject(ex.Content);
            string appId = jsonResponse["appId"].Value<string>();
            return appId;
        }
    }
}
```

I am using version "2.2.0-preview" of the NuGet package.

The inner exception thrown is as following:

```
InnerException: {"Unexpected character encountered while parsing value: {. Path 'appPermissions', line 1, position 427."}
    Message: "Unable to deserialize the response."
    Source: "Microsoft.Azure.Graph.RBAC"
    StackTrace: "   at Microsoft.Azure.Graph.RBAC.ApplicationOperations.<CreateWithHttpMessagesAsync>d__5.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at Microsoft.Azure.Graph.RBAC.ApplicationOperationsExtensions.<CreateAsync>d__1.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)\r\n   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()\r\n   at Authentication.Program.<CreateApplication1>d__10.MoveNext() 
```

I believe the problem lies in [Application.cs](https://github.com/Azure/azure-sdk-for-net/blob/master/src/Graph.RBAC/Graph.RBAC/Generated/Models/Application.cs). In this generated model class, [AppPermissions](https://github.com/Azure/azure-sdk-for-net/blob/9a460f87edc31408978f847ab931d4c54defaf07/src/Graph.RBAC/Graph.RBAC/Generated/Models/Application.cs#L50) is expected to be a list of string, but in reality it is a list of complex objects which looks like this:

``` json
  "appPermissions": [
    {
      "claimValue": "user_impersonation",
      "description": "Allow the application to access TestApp on behalf of the signed-in user.",
      "directAccessGrantTypes": [ ],
      "displayName": "Access TestApp",
      "impersonationAccessGrantTypes": [
        {
          "impersonated": "User",
          "impersonator": "Application"
        }
      ],
      "isDisabled": false,
      "origin": "Application",
      "permissionId": "XXX",
      "resourceScopeType": "Personal",
      "userConsentDescription": "Allow the application to access TestApp on your behalf.",
      "userConsentDisplayName": "Access TestApp",
      "lang": null
    }
  ]
```

How can the generated model be fixed?

**Comment by markcowl** [comment]
@shuagarw  can you take  a look?

**Comment by shuagarw** [comment]
+ @stefmilmsft
Well... I tried to repro, but looks like I don't get back any 'appPermissions' section. Will follow up with the graph team on when is it returned back.

**Comment by estiller** [comment]
Hi, I reproduced the issue now with the same code in the linked GitHub repo. Could it be settings related to a specific Tenant or Subscription?

**Comment by shuagarw** [comment]
Hi, Can you please use package version 1.9.0-preview (https://www.nuget.org/packages/Microsoft.Azure.Graph.RBAC/1.9.0-preview). 
1.9.0-preview is actually using the new graph version (1.6-internal). Try this and let me know if you see any issue.

**Comment by estiller** [comment]
Well, I downgraded the NuGet package from 2.2 to 1.9 as you suggested and got the program to compile, but now when I try to create the application, a Cloud Exception is thrown ("Hyak.Common.CloudException") whose message is:
"{"odata.error":{"code":"Authentication_MissingOrMalformed","message":{"lang":"en","value":"Access Token missing or malformed."},"values":null}}"

Something is now wrong with the authentication and I can't figure out what it is. I tried several options but all of them got the same result. You can look at the updated code in a branch I opened [here](https://github.com/estiller/GraphRBACException/tree/RBAC_v1.9). You can see the authentication options I tried [here](https://github.com/estiller/GraphRBACException/blob/RBAC_v1.9/GraphRBACException/Program.cs#L72).

**Comment by shuagarw** [comment]
@rashidqureshi @stefmilmsft 
Hi Eran, Earlier the sdk you were using had graph api version 1.42-previewInternal which could use the same token fetched for different resource (https://management.core.windows.net/).
Now, in new graph api versions, the resource while acquiring token should be the same in which token will be used. Please change code (https://github.com/estiller/GraphRBACException/blob/RBAC_v1.9/GraphRBACException/Program.cs#L32) to use graph resource "https://graph.windows.net".

**Comment by estiller** [comment]
Well, I changed the code as you suggested and not it works.
BUT - I opened up the HTTP response in Fiddler and the reason that it works is that there is no such property "appPermissions" returned now from this version of the API and so it is returned empty.
In the returned JSON object, there is now instead a property called "oauth2Permissions" which represents pretty much what "appPermissions" did before, but it is not mapped to any property in the returned object in the SDK so there is no apparent issue.
The bottom line is that while the code indeed does not crash now, something in the deserialization still does not match between the SDK and the API.

**Comment by shuagarw** [comment]
Yes, you are right. We are aware of that and will be fixed in next release. We have to make a bunch of modifications to graph rbac sdk and it will be a part of that.

**Comment by adrianhall** [comment]
It looks like the new version came out (and this is just old and needs to be closed).  Can you confirm your issue is corrected in later versions?

**Comment by estiller** [comment]
@adrianhall wow I completely forgot about this! Anyway, I upgraded to the latest NuGet versions and it seems to work OK now so you can close the issue.
I'm a bit surprised that 3 years later, the NuGet package "Microsoft.Azure.Graph.RBAC" is still in preview...
